### PR TITLE
Use type conversion in `SSHConnection` hashtables when value doesn't match expected type

### LINKED ISSUE #10687: Invoke-Command -SSHConnection breaks if the HostName string is [psobject]-wrapped
The root cause is that a string returned from a cmdlet call or external program has an invisible extra `[psobject]` wrapper, and the `Invoke-Command` cmdlet fails to take that possibility into account.

In short: The command below breaks, because `(hostname)` has such an invisible wrapper, as evidenced by `(hostname)  -is [psobject]` being `$true`.

This is yet another manifestation of our old friend #5579.

The specific problem in `Invoke-Command` is [here](https://github.com/PowerShell/PowerShell/blob/77ddec161417f98f8f2e57e9e9ef57d00cdc4c98/src/System.Management.Automation/engine/remoting/commands/PSRemotingCmdlet.cs#L1063).

# Steps to reproduce

```powershell
{ icm -SSHConnection @{ Hostname = (hostname) } -ScriptBlock { 'hi' }  } | 
  Should -not -Throw
```

# Expected behavior

The test should succeed, if the local computer is set up for being a SSH remoting target.

# Actual behavior

The test fails:

```
Expected no exception to be thrown, but an exception 
"The provided SSHConnection hashtable parameter name or element is null or empty." was thrown from line:1 char:3
```

Note that using `(hostname).psobject.BaseObject` or `"$(hostname)"` or `(hostname).ToString()` makes the problem go away.

As an aside: if the host name were truly invalid due not being a string, the error message would be misleading, because the problem is that the entry is of the _wrong type_, not that it is _null or empty_.

A further aside: The comment above the `SSHConnection` parameter definition is incomplete: keys `Port` and `Subsystem` are missing.

# Environment data

```none
PowerShell Core v7.0.0-preview.4
```

### PR DESCRIPTION
**Pull request by SeeminglyScience**
# PR Summary

Use `LanguagePrimitives.ConvertTo` when reading a hashtable provided to the `SSHConnection` parameter.

## PR Context

Currently if the hashtable value does not match the expected type exactly, an exception is thrown stating that the value was `null`.

Fixes #10687

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [x] N/A or can only be tested interactively
    - **OR**
    - [ ] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [x] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and enumerated concerns in the summary. This may include:
        - Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode (which runs in a different PS Host).
        - Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
        - Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
        - Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).

### CODE REVIEW

**Review by vexx32 — APPROVED** [review]

File: src/System.Management.Automation/engine/remoting/commands/PSRemotingCmdlet.cs [hunk 1]
```diff
@@ -1060,7 +1060,8 @@ protected void ValidateComputerName(string[] computerNames)
         /// <returns>Parameter value as string.</returns>
         private static string GetSSHConnectionStringParameter(object param)
         {
-            if (param is string paramValue && !string.IsNullOrEmpty(paramValue))
+            string paramValue = LanguagePrimitives.ConvertTo<string>(param);
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/remoting/commands/PSRemotingCmdlet.cs]: ConvertTo() can throw PSInvalidCastException but we should keep PSArgumentException.

File: src/System.Management.Automation/engine/remoting/commands/PSRemotingCmdlet.cs [hunk 2]
```diff
@@ -1075,12 +1076,12 @@ private static string GetSSHConnectionStringParameter(object param)
...
         {
-            if (param is int paramValue)
+            if (param == null)
             {
-                return paramValue;
+                throw new PSArgumentException(RemotingErrorIdStrings.InvalidSSHConnectionParameter);
             }
 
-            throw new PSArgumentException(RemotingErrorIdStrings.InvalidSSHConnectionParameter);
+            return LanguagePrimitives.ConvertTo<int>(param);
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/remoting/commands/PSRemotingCmdlet.cs]: The same.

**Review by iSazonov — COMMENTED** [review]
Please add tests.

**Comment by SeeminglyScience** [comment]
> Please add tests.

Any thoughts on that?  If I try to test that the conversion works then we'll hit the interactive SSH prompt.  I could give it a bad port and/or hostname, but that'll add ~5 seconds per test.  If that's acceptable I'll add a few of those.

File: src/System.Management.Automation/engine/remoting/commands/PSRemotingCmdlet.cs (hunk 1 above)
Comment (SeeminglyScience) [reply src/System.Management.Automation/engine/remoting/commands/PSRemotingCmdlet.cs]: Is there a circumstance where converting to `string` will throw?

Comment (iSazonov) [reply src/System.Management.Automation/engine/remoting/commands/PSRemotingCmdlet.cs]: Yes, it is not common but possible (if someone creates ToString() method which throw).

**Comment by iSazonov** [comment]
I'd expect 2 test for these two code paths modified.
Please use Feature and Slow tags in the tests.

**Comment by SeeminglyScience** [comment]
@iSazonov I've added tests and normalized the exception type.  In the tests I used port `49151` since it's reserved and shouldn't be in use, but that was the result of a quick search so I'm open to suggestions.

File: test/powershell/engine/Remoting/SSHRemotingCmdlets.Tests.ps1 [hunk 3]
```diff
@@ -40,3 +40,21 @@ Describe "SSHConnection parameter hashtable error conditions" -Tags 'Feature' {
...
+        # then New-PSSession will throw before even attempting to open the session.
+        $TestCasesSSHConnection = @(
+            @{scriptBlock = {New-PSSession -ErrorAction Stop -SSHConnection @{ Port = 49151; ComputerName = [psobject]'localhost' }}; testName = 'SSHConnection can convert PSObject to string'}
+            @{scriptBlock = {New-PSSession -ErrorAction Stop -SSHConnection @{ Port = [psobject]49151; ComputerName = 'localhost' }}; testName = 'SSHConnection can convert PSObject to int'}
+        )
+    }
+
+    It "<testName>" -TestCases $TestCasesSSHConnection {
+        param($scriptBlock)
+        { & $scriptBlock } | Should -Throw -ErrorId '2100,PSSessionOpenFailed'
```
Comment (iSazonov) [inline test/powershell/engine/Remoting/SSHRemotingCmdlets.Tests.ps1]: It is out of the PR but I'd prefer to see more useful message.

**Review by iSazonov — APPROVED** [review]

**Review by SteveL-MSFT — APPROVED** [review]

**Review by PaulHigin — APPROVED** [review]

**Comment by iSazonov** [comment]
@SeeminglyScience Thanks for your contribution!

**Pull request merged**